### PR TITLE
Removing Iterator and Page subclasses.

### PR DESCRIPTION
**Pull request by dhermes**
Instead require `Iterator` takes:
- a well-formed path for the request
- a callable to convert a JSON item to native obj.
- (optional) the key in a response holding all items
- (optional) a `page_start` (acts as proxy for `Page.__init__`)

### CODE REVIEW

**Comment by jonparrott** [comment]
@dhermes the general idea here LGTM.

**Comment by dhermes** [comment]
@tseaver PTAL. I'd really like to get the `list_foo()` migration done

**Comment by daspecster** [comment]
Ditto @jonparrott.

File: core/google/cloud/iterator.py [hunk 1]
```diff
@@ -175,12 +220,23 @@ def next(self):
...
+
+    :type items_key: str
+    :param items_key: The key used to grab retrieved items from an API
+                      response. Defaults to :data:`DEFAULT_ITEMS_KEY`.
+
+    :type item_to_value: callable
+    :param item_to_value: (Optional) Callable to convert an item from JSON
+                          into the native object. Assumed signature
+                          takes an :class:`Iterator` and a dictionary
+                          holding a single item.
```
Comment (tseaver) [inline core/google/cloud/iterator.py]: Why is this argument optional?  The iterator is pretty much useless without it, no?

File: core/google/cloud/iterator.py [hunk 2]
```diff
@@ -175,12 +220,23 @@ def next(self):
...
     :type client: :class:`~google.cloud.client.Client`
     :param client: The client, which owns a connection to make requests.
 
+    :type path: str
+    :param path: The path to query for the list of items. Defaults
+                 to :attr:`PATH` on the current iterator class.
+
+    :type items_key: str
+    :param items_key: The key used to grab retrieved items from an API
+                      response. Defaults to :data:`DEFAULT_ITEMS_KEY`.
```
Comment (tseaver) [inline core/google/cloud/iterator.py]: *This* one is optional.

Comment (dhermes) [inline core/google/cloud/iterator.py]: I take this to mean you'd like me to update the docstring to say "(Optional) The key..."?

**Comment by dhermes** [comment]
@tseaver Anything else?

File: core/google/cloud/iterator.py (hunk 1 above)
Comment (tseaver) [inline core/google/cloud/iterator.py]: But why not make it a required positional argument instead of faux-virtual?

Comment (dhermes) [inline core/google/cloud/iterator.py]: That's fine with me. Will do.

File: core/google/cloud/iterator.py (hunk 2 above)
Comment (tseaver) [inline core/google/cloud/iterator.py]: Yes, since there is a "normal" default, the parameter is optional.

**Comment by dhermes** [comment]
@tseaver So I'll add "(Optional)" to the docstring and make `item_to_value` a positional / required argument.

Good to merge after those changes are in?

**Comment by tseaver** [comment]
@dhermes 

> Good to merge after those changes are in?

Yup.

**Comment by dhermes** [comment]
OK, waited an hour for Travis. Merging. (There is a lot to do after.)

**Pull request merged**